Add failOnFlakeCount option

oehme opened the pull request:
To clarify https://issues.apache.org/jira/browse/SUREFIRE-1878 for further discussion.

paplorinc commented on maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/VerifyMojo.java:
```diff
@@ -139,6 +139,12 @@
+     * The number of flakes after which the overall test suite will be considered a failure.
+     */
+    @Parameter( property = "surefire.failOnFlakeCount", defaultValue = "0" )
+    private int failOnFlakeCount;
```
Could we clarify if this is total number of flakes per build (is that what "overall" refers to?) or per test (i.e. two test methods are both retried 3 times, does the build fail for `failOnFlakeCount = 3` or `failOnFlakeCount = 6`)?

Tibor17 replied:
Exactly this was not clear to me when we spoke with the JUnit5 colleagues several days ago.
The logic of `re-run` feature operates at the level of individual Test.
In this PR I guess we are talking about the level of Test Set (cumulative number of flakes) which is different view from ours. But's that's ok if it would be properly clarified.

paplorinc replied:
It's very clear now, thanks!

paplorinc commented on maven-surefire-plugin/src/main/java/org/apache/maven/plugin/surefire/SurefirePlugin.java:
```diff
@@ -253,6 +253,12 @@
+    /**
+     * The number of flakes after which the overall test suite will be considered a failure.
+     */
+    @Parameter( property = "surefire.failOnFlakeCount", defaultValue = "0" )
```
Could we specify here whether `0` means a single retry will fail the build (i.e. similar to disabling retries completely, I guess), or whether this disables the feature (which seems to be the case based on https://github.com/apache/maven-surefire/pull/333/files#diff-bf7f174589af8f08638da17d6aefbe897133b7b77ff363a1e12439c37ea8683eR146, which seems weird to me, maybe we'd have to come up with a better name)

paplorinc commented on surefire-its/src/test/java/org/apache/maven/surefire/its/JUnitPlatformRerunFailingTestsIT.java:
```diff
@@ -203,6 +203,19 @@ public void testRerunOneTestMethod()
+    {
+        OutputValidator outputValidator = unpack().setJUnitVersion( VERSION ).maven().debugLogging()
+            .addGoal( "-Dsurefire.rerunFailingTestsCount=2" )
+            .addGoal( "-Dsurefire.failOnFlakeCount=1" )
```
could we add a test for `rerunFailingTestsCount=0` also - or is that tested implicitly, being the default?

Tibor17 replied:
Then it would become a simple test and it should perform as if `failOnFlakeCount` was disabled. Am I right?
So the `failOnFlakeCount` works only with specified `rerunFailingTestsCount > 0`?

paplorinc commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireHelper.java:
```diff
@@ -286,7 +288,8 @@ private static String createErrorMessage( SurefireReportParameters reportParamet
         {
-            msg.append( "There are test failures.\n\nPlease refer to " )
+            msg.append( result.getFailures() > 0 ? "There are test failures." : "There are too many flakes." );
+            msg.append( "\n\nPlease refer to " )
```
minor: maybe we could merge these appends, as done in the next few lines

Tibor17 commented:
@oehme 
Thx, that's nice but I have a technical problem I described to your colleague in the Slack. The problem is the ID of the runs. It's still the same and Surefire cannot modify it. That's the reason why the re-run works with same ID of test run. Once you set `failOnFlakeCount` we would have to identify these test runs as different ones. I want to say that these would become ordinal tests in the XML report.
The next issue with understanding `failOnFlakeCount` is the count as `integer`. I would understand to have `boolean` but the `integer` is strange because nobody has explained how to handle this number and how to mark these tests.

Tibor17 commented:
@oehme 
Do you want to have test set resolution for entire suite?
`result.getFlakes() < reportParameters.getFailOnFlakeCount()`
We understood the Jira issue that you want to have `boolean` related to individual Test and not test set.
Pls clarify and change the description in Jira ticket as necessary.

oehme commented:
It's about the whole test set. I don't understand the point about IDs. I'm not changing anything about test IDs, I'm just adding validation at the end of the execution, similar to failOnError.

Tibor17 commented:
@oehme 
The IDs because it was related to what I have said that our understanding of rerun was at the level of individual test. If we want to have only a cumulative counter than it is different story of course.

oehme commented:
Got it, thanks! 

So is this option something you'd consider adding? In that case I'd flesh out the PR some more.

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireHelper.java:
```diff
@@ -142,7 +142,9 @@ public static void reportExecution( SurefireReportParameters reportParameters, R
         throws MojoFailureException, MojoExecutionException
     {
-        if ( firstForkException == null && !result.isTimeout() && result.isErrorFree() )
+        if ( firstForkException == null && !result.isTimeout() && result.isErrorFree()
```
@oehme 

My proposal:
```
boolean isError = firstForkException != null || result.isTimeout() || !result.isErrorFree();
boolean isFlaky = reportParameters.getFailOnFlakeCount() > 0
                && result.getFlakes() >= reportParameters.getFailOnFlakeCount();
if ( isError || isFlaky )
{
    ...
}
else
{
    if ( isError )
    {
        msg.append( "There are test failures." );
    }

    if ( isFlaky )
    {
        msg.append( "There are too many flakes." );
    }
    
    msg.append( "\n\nPlease refer to " );
}
```

Tibor17 replied:
> 
> 
> Got it, thanks!
> 
> So is this option something you'd consider adding? In that case I'd flesh out the PR some more.

Yes I would consider this PR adding to Surefire.
Scary of the future PRs a bit ;-)

Tibor17 commented:
@oehme 
I think the class `SurefireHelper.java` has unit tests. Try to cover every change you made. Thx

oehme commented:
I've clarified the documentation and added unit tests. I've removed the integration test, since this feature is not junit-platform specific, but fully contained in SurefireHelper.

Are there any other things I should add or modify?

paplorinc commented on maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/VerifyMojo.java:
```diff
@@ -139,6 +139,13 @@
 
+    /**
+     * Set this to a value greater than 0 to fail the whole test set if the cumulative number of flakes reaches
+     * this threshold. Set to 0 to allow an unlimited number of flakes.
```
based on the code a negative number would do the same, right?
Should we maybe rather fail in that case?

Tibor17 replied:
@oehme
Actually yes, we should and there is something like "verifyParameters" method.
Have you used the flag to serialize the content in the method `getConfigChecksum`?

paplorinc commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireHelper.java:
```diff
@@ -142,7 +142,10 @@ public static void reportExecution( SurefireReportParameters reportParameters, R
-        if ( firstForkException == null && !result.isTimeout() && result.isErrorFree() )
+        boolean isError = firstForkException != null || result.isTimeout() || !result.isErrorFree();
+        boolean isFlaky = reportParameters.getFailOnFlakeCount() > 0
+            && result.getFlakes() >= reportParameters.getFailOnFlakeCount();
```
minor: could we have `getFailOnFlakeCount` on the same side on both cases to simplify comparison?

Tibor17 replied:
@paplorinc 
What you mean?

paplorinc replied:
Just an observation that `X > 0 && X < Y` may be easier to read than `X > 0 && Y >= X`.
Don't have strong feelings about it, just an observation

Tibor17 replied:
@paplorinc 
I think `result.getFlakes() >= reportParameters.getFailOnFlakeCount()` is okay because it is a condition saying `variable < constant`.

paplorinc commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireHelper.java:
```diff
@@ -142,7 +142,10 @@ public static void reportExecution( SurefireReportParameters reportParameters, R
     {
-        if ( firstForkException == null && !result.isTimeout() && result.isErrorFree() )
+        boolean isError = firstForkException != null || result.isTimeout() || !result.isErrorFree();
+        boolean isFlaky = reportParameters.getFailOnFlakeCount() > 0
```
this stores more than whether we're flaky, rather something like: `isFlakyError`, right?

Tibor17 replied:
minor i would say

paplorinc commented on maven-surefire-common/src/test/java/org/apache/maven/plugin/surefire/SurefireHelperTest.java:
```diff
@@ -156,4 +156,59 @@ public void shouldHandleTestFailure() throws Exception
+                    + "[date].dump, [date]-jvmRun[N].dump and [date].dumpstream." );
+            return;
+        }
+        fail( "Expected MojoFailureException with message "
```
minor: you could add this inside the try in which case we wouldn't need a return in the catch

Tibor17 replied:
@oehme 
Pls move this line right after the line 168, and remove the `return`. It would behave the same, We normally do it this way.

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireHelper.java:
```diff
@@ -286,7 +289,15 @@ private static String createErrorMessage( SurefireReportParameters reportParamet
         else
         {
-            msg.append( "There are test failures.\n\nPlease refer to " )
+            if ( result.getFailures() > 0 )
```
@oehme 
Pls see the class `IntegrationTestPlugin`. It reports this text other way. The same messages should be there as well but the solution is different from this one.

oehme replied:
The IntegrationTestMojo does not report this text. The VerifyMojo does and uses the same logic as the SurefirePlugin

Tibor17 replied:
@oehme 
ok just fine!

Tibor17 commented:
> 
> 
> I would have used that normally, but all the other methods in this test didn't use it either, but instead used the pattern shown here. I generally always go with the pattern that some codebase already uses.
> 
> Happy to use ExpectedException. Should I also adjust the other existing methods in this test?

Pls first use `ExpectedException` for your two methods and then squash your commits into one.
Then concentrate on the `ExpectedException` for the other methods and make extra commit.
+1 LGTM

oehme commented:
All fixed up

Tibor17 commented:
> 
> 
> All fixed up

Thx. Waiting for the CI.

Tibor17 commented on maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/VerifyMojo.java:
```diff
@@ -139,6 +139,13 @@
+     * Set this to a value greater than 0 to fail the whole test set if the cumulative number of flakes reaches
+     * this threshold. Set to 0 to allow an unlimited number of flakes.
+     */
+    @Parameter( property = "surefire.failOnFlakeCount", defaultValue = "0" )
```
Here the property `surefire.failOnFlakeCount` must become `failsafe.failOnFlakeCount`. See the other properties in this class.

Tibor17 commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireHelper.java:
```diff
@@ -286,7 +289,15 @@ private static String createErrorMessage( SurefireReportParameters reportParamet
+            }
+            else
+            {
+                msg.append( "There are too many flakes." );
```
@oehme 
I think this would not be printeted very often even if it could because there is if-else.
We do not have IT test but both messages could be printed along. Why the message `There are too many flakes.` does not have the concrete numbers. We have both numbers available and they are the cause why the message is going to be printed. I only want to give the user all reason why the plugin crashed.

oehme replied:
Done, both messages are reported now if both conditions are true. Also, the number of flakes and failOnFlakeCount are reported now.

Tibor17 replied:
@oehme 
Thx, a good job!

Tibor17 commented on maven-surefire-plugin/src/main/java/org/apache/maven/plugin/surefire/SurefirePlugin.java:
```diff
@@ -880,4 +900,20 @@ protected final ForkNodeFactory getForkNode()
     }
+
+    @Override
+    boolean verifyParameters() throws MojoFailureException, MojoExecutionException
```
This must not exist in both Plugin classes. See the super method in AbstractSurefireMojo and add a new warn* method and a call which finally calls the abstract method getFailOnFlakeCount.

oehme replied:
SurefirePlugin and VerifyMojo do not have a common base class.

Tibor17 replied:
I am not talking about VerifyMojo, nothing but the SurefirePlugin class.

oehme replied:
But why would the super class contain logic that only applies to a single subclass?

oehme replied:
Nevermind, I see the if-else in that method now.

Tibor17 replied:
> 
> 
> But why would the super class contain logic that only applies to a single subclass?

Both plugins, Surefire and Failsafe, have 90% same parameters. So they always contained the method `verifyParameters()` in the super class. The `VerifyMojo` is an exception and it has never had so many config params.

Tibor17 commented on maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/VerifyMojo.java:
```diff
@@ -236,6 +243,11 @@ protected boolean verifyParameters()
             return false;
         }
 
+        if ( failOnFlakeCount < 0 )
```
This must be in the abstract class where are all the warning check method calls made in the section:
```
        else
        {
            ensureEnableProcessChecker();
            convertDeprecatedForkMode();
            ensureWorkingDirectoryExists();
            ensureParallelRunningCompatibility();
            ensureThreadCountWithPerThread();
            warnIfUselessUseSystemClassLoaderParameter();
            warnIfDefunctGroupsCombinations();
            warnIfRerunClashes();
            warnIfWrongShutdownValue();
            warnIfNotApplicableSkipAfterFailureCount();
            warnIfIllegalTempDir();
            warnIfForkCountIsZero();
            printDefaultSeedIfNecessary();
        }
```

Tibor17 commented on maven-surefire-plugin/src/main/java/org/apache/maven/plugin/surefire/SurefirePlugin.java:
```diff
@@ -880,4 +900,20 @@ protected final ForkNodeFactory getForkNode()
+    @Override
+    boolean verifyParameters() throws MojoFailureException, MojoExecutionException
+    {
+        if ( failOnFlakeCount < 0 )
```
As in my previous comment, this must be in the abstract class where are all the warning check method calls made in the section:
```
        else
        {
            ensureEnableProcessChecker();
            convertDeprecatedForkMode();
            ensureWorkingDirectoryExists();
            ensureParallelRunningCompatibility();
            ensureThreadCountWithPerThread();
            warnIfUselessUseSystemClassLoaderParameter();
            warnIfDefunctGroupsCombinations();
            warnIfRerunClashes();
            warnIfWrongShutdownValue();
            warnIfNotApplicableSkipAfterFailureCount();
            warnIfIllegalTempDir();
            warnIfForkCountIsZero();
            printDefaultSeedIfNecessary();
        }
```

oehme replied:
See above, this wouldn't make sense because IntegrationTestMojo does not have `failOnFlakeCount`. That's in the VerifyMojo, which doesn't have the same base class.

Tibor17 replied:
@oehme 
I am not talking about VerifyMojo, nothing but the SurefirePlugin class.

Tibor17 commented:
@oehme 
Do you have somewhere a backup of the initial IT test?

@eolivelli 
I am glad of this code. WDYT, can we push it to master?

eolivelli reviewed (approved):
LGTM

sure, I like this feature

thank you @oehme 

@Tibor17 can out please push to an ASF repo branch in order to see CI validating the patch ?

oehme commented:
@Tibor17 I don't, but I can easily add one back if you want. I removed it because this feature isn't specific to a particular provider and all the logic could be unit tested, so I didn't feel like an extra integ test was necessary anymore.

eolivelli commented:
@oehme I think there is not need to add the IT

@Tibor17  ?

Tibor17 commented:
I also believe that it would be fine without IT but we should cover every config param by IT. It would be the highest guarantee that the client would not fire a bug against us that a fresh feature is dead ;-)

oehme commented:
Added the IT back

Tibor17 commented:
@oehme 
@eolivelli 
Thx for contributing!

Tibor17 commented:
@oehme 
@eolivelli 
One more thing I have noticed which could be made better is to improve the condition in method `warnIfIllegalFailOnFlakeCount()` because this should fire an exception in a condition with setting `rerunFailingTestsCount`. The documentation of rerun should mention a new config param. This was also forgotten.

oehme commented:
If you want I can send another PR, otherwise I'll let you take care of it.

Tibor17 commented:
@oehme 
Hi, that would be great, thank you.

oehme commented:
I've looked at this and I think it would be inconsistent: Only the SurefirePlugin has both `failOnFlakeCount` and `rerunFailingTestCount`. For integration tests, the two options are split between `IntegrationTestMojo` and `VerifyMojo`. So my suggestion would be to leave the documentation/validation as-is. Unless I'm missing an option for cross-mojo validation/documentation.

Tibor17 commented:
We have two plugins: surefire and failsafe.
The one you mentioned `VerifyMojo` is still the same failsafe plugin but a separate goal.
It normal practice to add one sentecnce to the existing [documentation](https://github.com/apache/maven-surefire/blob/master/maven-surefire-plugin/src/site/apt/examples/rerun-failing-tests.apt.vm#L150). Last time we added `skipAfterFailureCount` to the re-run. Here we should add a next chapter `Re-run and fail the build upon flaky test count`. That's all.

oehme commented:
I was referring to the additional validation you wanted me to add - That's possible for surefire (both properties in the same Mojo), but as far as I can see not for failsafe (properties spread over two Mojos)

Tibor17 commented:
@oehme 
We have some old JIRA issue about improving the validations reported by @krosenvold . This can be reused and fixed separately, aside of this issue.
